### ISSUE TITLE
When entering insert mode again from normal mode in repl , the backspace can't delete character.

### ISSUE THREAD

**Issue by pittcat** [mfussenegger/nvim-dap#291]
### Debug adapter definition and debug configuration

[nvim-dap-python](https://github.com/mfussenegger/nvim-dap-python)

### Debug adapter version

latest

### Steps to Reproduce

1. Input any character in repl buffer
2. Press <Esc> to leave insert mode
3. Enter insert mode again
4. Press <Backspace> to delete character

### Expected Result

Backspace can delete character when entering insert mode again by some workaround.

### Actual Result

https://asciinema.org/a/oQgELi5G8rkCvUfqLNnSHnuty

**Comment by mfussenegger** [comment]
As far as I'm aware this is is a vim/neovim issue with the prompt buftype. I attempted workarounds, but so far failed to come up with anything that is reliable.

This needs to be fixed upstream.

**Comment by wookayin** [comment]
@mfussenegger Do you have any pointer to the upstream issue? Thanks!

**Comment by mfussenegger** [comment]
@wookayin I think https://github.com/neovim/neovim/pull/15881 might be related, but I didn't verify.

**Comment by pfeiferj** [comment]
@pittcat I put my workaround for this in #119 ([comment](https://github.com/mfussenegger/nvim-dap/issues/119#issuecomment-976046944)) not realizing there was an open issue for it. I also created #357 to potentially resolve this but not sure if it's something @mfussenegger would really want to merge in because it's not very clean.

**Comment by pfeiferj** [comment]
An improvement to the previously mentioned backspace workaround that doesn't switch between normal and insert modes:

```lua
function PromptBackspace()
  -- Allows backspacing through previously set text when in a prompt.
  --
  -- Note 1: Insert mode cursor is after (+1) the column as opposed to in normal mode it would be on (+0) the column.
  -- Note 2: nvim_win_[get|set]_cursor is (1, 0) indexed for (line, column) while nvim_buf_[get|set]_[lines|text] is 0 indexed for both.

  local currentCursor = vim.api.nvim_win_get_cursor(0)
  local currentLineNumber = currentCursor[1]
  local currentColumnNumber = currentCursor[2]
  local currentLine = vim.api.nvim_buf_get_lines(0, currentLineNumber-1, currentLineNumber, false)[1]
  local _, endPromptPrefix = string.find(currentLine, "> ")
 
  if endPromptPrefix == nil then
    endPromptPrefix = 0
  end

  if (currentColumnNumber) ~= endPromptPrefix then
    vim.api.nvim_buf_set_text(0, currentLineNumber-1, currentColumnNumber-1, currentLineNumber-1, currentColumnNumber, {""})
    vim.api.nvim_win_set_cursor(0, {currentLineNumber, currentColumnNumber-1})
  end
end

vim.cmd([[
fun! PromptBackspaceSetup()
    if v:option_new == 'prompt'
      inoremap <buffer> <backspace> <Cmd>lua PromptBackspace()<cr>
    endif
endfun

autocmd OptionSet * call PromptBackspaceSetup() "not using a lua function here because of error when accessing :help
]])
```

**Comment by mfussenegger** [comment]
https://github.com/neovim/neovim/pull/16421 might be related

**Comment by mfussenegger** [comment]
Can anyone confirm if this is fixed if using latest neovim master?

**Comment by pfeiferj** [comment]
Just got a chance to try it out and can confirm backspace is working properly on the latest neovim nightly: NVIM v0.7.0-dev+797-gb218d02c4

**Comment by mfussenegger** [comment]
Closing since it is fixed in neovim nightly
